### PR TITLE
[PM-1426] Refactor uri matching

### PR DESCRIPTION
**Pull request by eliykat**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [x] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Follow-up from PM-1397 to tidy up some duplicated logic and add unit tests.

## Code changes

**Primary changes**
- **libs/common/src/vault/models/view/login-uri.view.ts** - move the URI matching logic into this model to reduce duplication. 
  - reduce nested conditional logic in the switch statement to make Codescene happy
  - add unit tests
- **libs/common/src/vault/models/view/login.view.ts** - add a `uriMatches` method here as well, which is a convenience method to check all URIs on an item
- **apps/browser/src/autofill/services/autofill.service.ts** - use the new `matchesUri` method instead of duplicating code
  - also move the `pageUrl === tabUrl` check inside the `inUntrustedIframe` method to better encapsulate it.
- **libs/common/src/vault/services/cipher.service.ts** - use the new `matchesUri` method instead of duplicating code
- **libs/common/src/services/settings.service.ts** 
  - use a `Set` instead of an array to represent equivalent domains
  - add unit tests for `getEquivalentDomains`

**Secondary changes**
- **libs/common/src/misc/utils.ts** 
  - fix an edge case I found in `Utils.getUrl` where omitting the protocol but including a port number (e.g. `Utils.getUrl('www.example.com:4000')`) will return an incorrectly parsed `URL` object. It parses everything before the `:` as the protocol, which then means we can't extract parts of the URL for comparison. We already have logic which prepends `http` as the default protocol if none is present, but only if the `URL()` constructor throws first (presumably due to an invalid URL). However, this is a case where the URL constructor does **not** throw, and instead returns a result that is incorrect. I've tweaked this so that we check for the protocol _before_ calling the URL constructor, and add `http` as the default protocol if required. I also deleted some duplicate logic
  - if none of this makes sense, try this in your browser JS console and compare the results: `new URL('http://www.example.com:4000')` vs. `new URL('www.example.com:4000')`. Note how the protocol and host are different between the two results
  - remove old fallback method of parsing URLs, as `new URL` is supported in all modern browsers: https://caniuse.com/?search=URL
  - requesting a review from @djsmith85 on this part in particular as he's spent a lot of time on URL parsing

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

File: libs/common/src/vault/models/view/login-uri.view.ts
Comment (eliykat) [inline libs/common/src/vault/models/view/login-uri.view.ts:1]: We can take care of Codescene complaints after unit tests are filled out.

File: libs/common/src/vault/services/cipher.service.ts [hunk 1]
```diff
@@ -441,59 +408,8 @@ export class CipherService implements CipherServiceAbstraction {
...
-                this.logService.error(e);
-              }
-              break;
-            case UriMatchType.Never:
-            default:
-              break;
-          }
-        }
+      if (cipher.type === CipherType.Login && cipher.login !== null) {
+        return cipher.login.matchesUri(url, equivalentDomains, defaultMatch);
```
Comment (differsthecat) [inline libs/common/src/vault/services/cipher.service.ts:412]: So much better 💯

**Review by differsthecat — DISMISSED** [review]
This is great, thank you @eliykat !

**Review by djsmith85 — DISMISSED** [review]
@eliykat: Awesome work on this.

Dragging logService around just for the Regex case seemed like a waste anyway.

Changes to Utils look great and thanks for adding tests.

**Review by vincentsalucci — APPROVED** [review]
Approving because the dismissal was from merging master and updating import paths. @differsthecat and @djsmith85 did all the real reviewing!
